GAWB-1495: APIs for listing realms (admin and user)

MatthewBemis opened the pull request:
This is a PR into the branch created for GAWB-1494

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

MatthewBemis committed:
initial implementation

MatthewBemis committed:
use realm refs instead of group refs

MatthewBemis committed:
clarify swagger

MatthewBemis committed:
tests working

MatthewBemis committed:
initial basic tests for the api endpoints

MatthewBemis committed:
more tests

MatthewBemis committed:
don't need this no more

MatthewBemis committed:
block creating workspaces with regular groups as realms

MatthewBemis committed:
more tests

MatthewBemis committed:
remove comment from copypaste

MatthewBemis committed:
it's a string...

MatthewBemis committed:
dont need apply here anymore

dvoet commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala:
```diff
@@ -199,6 +199,14 @@ trait RawlsGroupComponent {
+    def listRealmsForUser(userRef: RawlsUserRef): ReadAction[Set[RawlsRealmRef]] = {
+      listAllRealms().flatMap { allRealms =>
+        listGroupsForUser(userRef).map { allGroups =>
+          allRealms intersect allGroups.map(ref => RawlsRealmRef(ref.name))
```
I think it is more correct to map realms to a group than a group to a realm.

jmthibault79 replied:
How does this compare to using a join instead?

dvoet reviewed (changes requested):
I don't think renaming GroupRef.groupName is right. I think it hints at a json marshaling issue.

MatthewBemis committed:
rollback previous 2 commits, mostly fixed stupid spray json implicit errors

MatthewBemis committed:
incorporate fix for realm comparison

MatthewBemis commented on core/src/main/resources/swagger/rawls.yaml:
```diff
@@ -675,80 +675,104 @@ paths:
+      parameters:
+        - in: body
+          description: group reference
+          name: groupRef
```
all of these changes are just whitespace except for the GET at the bottom

MatthewBemis commented on core/src/main/scala/org/broadinstitute/dsde/rawls/model/UserAuth.scala:
```diff
@@ -13,7 +14,7 @@ case class RawlsUserInfoList(userInfoList: Seq[RawlsUserInfo])
 
 object RawlsUser {
-  implicit def toRef(u: RawlsUser) = RawlsUserRef(u.userSubjectId)
+  implicit def toRef(u: RawlsUser): RawlsUserRef = RawlsUserRef(u.userSubjectId)
```
added explicit implicit return types while i was here

jmthibault79 replied:
yay

MatthewBemis commented on model/src/main/scala/org/broadinstitute/dsde/rawls/model/JsonSupport.scala:
```diff
@@ -95,7 +96,7 @@ trait TypedAttributeListSerializer extends AttributeListSerializer {
  * NOTE: When subclassing this, you may want to define your own implicit val attributeFormat with an AttributeListSerializer mixin
  * if you want e.g. the plain old JSON array attribute list serialization strategy
  */
-class JsonSupport extends DefaultJsonProtocol {
```
Hussein's investigation into spray json implicit errors rescued me from this evil line

i was getting all sorts of spray json implicit errors even when importing the unmarshallers right above where they were being used. removing the `extends DefaultJsonProtocol` here fixed my problem but required me to import `DefaultJsonProtocol._` wherever we used spray json implicit unmarhsallers

MatthewBemis committed:
cleanup, stop mixing up realmName vs groupName terminology, etc

jmthibault79 commented on core/src/main/resources/swagger/rawls.yaml:
```diff
@@ -675,80 +675,104 @@ paths:
+    get:
+      responses:
+        '200':
+          description: Realm listing successful
```
add schema

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:
```diff
@@ -365,18 +365,18 @@ trait WorkspaceComponent {
-          val groupRef = RawlsGroupRef(RawlsGroupName(realm))
-          rawlsGroupQuery.loadGroupIfMember(groupRef, user) flatMap {
+          val realmRef = RawlsRealmRef(RawlsGroupName(realm))
+          rawlsGroupQuery.loadGroupIfMember(realmRef, user) flatMap {
```
`loadGroupIfMember` can be easily extended to take multiple groups, and would make the `DBIO.sequence()` unnecessary.

MatthewBemis replied:
Wouldn't this just push the `DBIO.sequence()` further down?

jmthibault79 replied:
No, there's no need to do a `sequence()` because you're getting all groups at once.  Then you can put the resulting `Seq[RawlsRealmRef]` inside a `DBIO.successful()`

jmthibault79 replied:
Never mind.  This is more complicated than I thought.

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponentSpec.scala:
```diff
@@ -311,6 +311,28 @@ class RawlsGroupComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers
+      runAndWait(rawlsGroupQuery.load(realm))
+    }
+
+    runAndWait(rawlsGroupQuery.listAllRealms()) should contain(RawlsRealmRef(realm.groupName))
```
Add a "should not contain" after Group save and before Set As Realm

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponentSpec.scala:
```diff
@@ -311,6 +311,28 @@ class RawlsGroupComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers
+    assertResult(0) {
+      runAndWait(rawlsGroupQuery.setGroupAsRealm(realm))
+    }
+  }
```
Add a "should contain"

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/AdminApiServiceSpec.scala:
```diff
@@ -421,6 +421,110 @@ class AdminApiServiceSpec extends ApiServiceSpec {
+    Get(s"/admin/realms") ~>
+      sealRoute(services.adminRoutes) ~>
+      check {
+        responseAs[Seq[RawlsRealmRef]] should contain(group)
```
Add a "should not contain" above

jmthibault79 commented on core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/AdminApiServiceSpec.scala:
```diff
@@ -421,6 +421,110 @@ class AdminApiServiceSpec extends ApiServiceSpec {
+    Get(s"/admin/realms") ~>
+      sealRoute(services.adminRoutes) ~>
+      check {
+        responseAs[Seq[RawlsRealmRef]] should not contain(group)
```
Add a "should contain" above

jmthibault79 commented:
Review done

MatthewBemis committed:
use toUserGroupRef instead of implicits, fix the billions of broken tests because of it

MatthewBemis committed:
remove commented-out implicits

MatthewBemis committed:
revert defaultjsonprotocol changes

MatthewBemis committed:
improve tests based on PR comments

MatthewBemis committed:
update swagger

MatthewBemis commented:
@jmthibault79 @dvoet 

back to you guys for re-review. I made all of the suggested changes unless I left a comment responding to it

I removed all implicits for converting between RawlsRealmRefs <-> RawlsGroupRefs and RawlsGroup -> RawlsRealmRef

I also removed the DefaultJsonProtocol changes that I made because Hussein is working on a more permanent fix for that and I didn't want to interfere w/ it

MatthewBemis committed:
remove one more import of DefaultJsonProtocol

jmthibault79 commented on core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala:
```diff
@@ -445,7 +449,7 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
   }
 
-  @deprecated
+  @deprecated(message = "this api is used in the old process of creating billing projects outside of firecloud, startBillingProjectCreation is the new way", since = "d9209db")
```
👍

jmthibault79 commented:
Good with one remaining comment.  If it's not clear, let me know if you want me to try updating the code with what I mean.

MatthewBemis committed:
small swagger changes (groupName -> realmName)

jmthibault79 commented:
👍

dvoet reviewed (approved):
👍
